R-package/src/lightgbm_R.cpp

```diff
@@ -77,97 +80,111 @@ SEXP LGBM_DatasetCreateFromCSC_R(SEXP indptr,
   SEXP num_row,
   SEXP parameters,
   SEXP reference) {
+  int n_protected = 0;
   SEXP ret;
-  R_API_BEGIN();
   const int* p_indptr = INTEGER(indptr);
   const int* p_indices = INTEGER(indices);
   const double* p_data = REAL(data);
-
   int64_t nindptr = static_cast<int64_t>(Rf_asInteger(num_indptr));
   int64_t ndata = static_cast<int64_t>(Rf_asInteger(nelem));
   int64_t nrow = static_cast<int64_t>(Rf_asInteger(num_row));
+  const char* parameters_ptr = CHAR(PROTECT(Rf_asChar(parameters)));
+  n_protected++;
   DatasetHandle handle = nullptr;
   DatasetHandle ref = nullptr;
   if (!Rf_isNull(reference)) {
     ref = R_ExternalPtrAddr(reference);
   }
+  R_API_BEGIN();
   CHECK_CALL(LGBM_DatasetCreateFromCSC(p_indptr, C_API_DTYPE_INT32, p_indices,
     p_data, C_API_DTYPE_FLOAT64, nindptr, ndata,
-    nrow, CHAR(Rf_asChar(parameters)), ref, &handle));
+    nrow, parameters_ptr, ref, &handle));
+  R_API_END();
   ret = PROTECT(R_MakeExternalPtr(handle, R_NilValue, R_NilValue));
+  n_protected++;
   R_RegisterCFinalizerEx(ret, _DatasetFinalizer, TRUE);
-  UNPROTECT(1);
+  UNPROTECT(n_protected);
   return ret;
-  R_API_END();
 }
 
 SEXP LGBM_DatasetCreateFromMat_R(SEXP data,
   SEXP num_row,
   SEXP num_col,
   SEXP parameters,
   SEXP reference) {
+  int n_protected = 0;
   SEXP ret;
-  R_API_BEGIN();
   int32_t nrow = static_cast<int32_t>(Rf_asInteger(num_row));
   int32_t ncol = static_cast<int32_t>(Rf_asInteger(num_col));
   double* p_mat = REAL(data);
+  const char* parameters_ptr = CHAR(PROTECT(Rf_asChar(parameters)));
+  n_protected++;
   DatasetHandle handle = nullptr;
   DatasetHandle ref = nullptr;
   if (!Rf_isNull(reference)) {
     ref = R_ExternalPtrAddr(reference);
   }
+  R_API_BEGIN();
   CHECK_CALL(LGBM_DatasetCreateFromMat(p_mat, C_API_DTYPE_FLOAT64, nrow, ncol, COL_MAJOR,
-    CHAR(Rf_asChar(parameters)), ref, &handle));
+    parameters_ptr, ref, &handle));
+  R_API_END();
   ret = PROTECT(R_MakeExternalPtr(handle, R_NilValue, R_NilValue));
+  n_protected++;
   R_RegisterCFinalizerEx(ret, _DatasetFinalizer, TRUE);
-  UNPROTECT(1);
+  UNPROTECT(n_protected);
   return ret;
-  R_API_END();
 }
 
 SEXP LGBM_DatasetGetSubset_R(SEXP handle,
   SEXP used_row_indices,
   SEXP len_used_row_indices,
   SEXP parameters) {
+  int n_protected = 0;
   SEXP ret;
-  R_API_BEGIN();
   int32_t len = static_cast<int32_t>(Rf_asInteger(len_used_row_indices));
   std::vector<int32_t> idxvec(len);
-  // convert from one-based to zero-based index
+  // convert from one-based to  zero-based index
```

Comment:
```suggestion
  // convert from one-based to zero-based index
```

This extra space looks like a mistake, could you please revert it?